File: python/2751.py
class Solution:
    def survivedRobotsHealths(self, positions: list[int], healths: list[int], directions: str) -> list[int]:
        n = len(positions)
        robots = [(positions[i], i) for i in range(n)]
        robots.sort()
        
        stack = []
        for i in range(n):
            idx = robots[i][1]
            if directions[idx] == 'R':
                stack.append(idx)
            else:
                while stack:
                    if healths[stack[-1]] > healths[idx]:
                        healths[stack[-1]] -= 1
                        healths[idx] = -1
                        break
                    elif healths[stack[-1]] < healths[idx]:
                        healths[idx] -= 1
                        healths[stack.pop()] = -1
                    else:
                        healths[idx] = -1
                        healths[stack.pop()] = -1
                        break
        result = []
        for i in range(n):
            if healths[i] > 0:
                result.append(healths[i])
        return result

if __name__ == '__main__':
    s = Solution()
    positions = list(map(int, input().strip().split()))
    healths = list(map(int, input().strip().split()))
    directions = input()
    print(s.survivedRobotsHealths(positions, healths, directions))

File: cpp/2751.cpp
#include<iostream>
#include<algorithm>
#include<vector>
#include<string>

using namespace std;

class Solution {
public:
    vector<int> survivedRobotsHealths(vector<int>& positions, vector<int>& healths, string directions) {
        int n = positions.size();
        vector<pair<int, int>> robots(n);
        for (int i = 0; i < n; ++i) {
            robots[i] = {positions[i], i};  // store position and index
        }
        
        // Sort robots based on positions
        sort(robots.begin(), robots.end());
        
        vector<int> stack;  // stack to store indices of robots moving to the right
        
        for (int i = 0; i < n; ++i) {
            int idx = robots[i].second;
            if (directions[idx] == 'R') {
                stack.push_back(idx);
            } else {
                while (!stack.empty()) {
                    int last = stack.back();
                    if (healths[last] > healths[idx]) {
                        healths[last] -= 1;
                        healths[idx] = -1;  // current robot is destroyed
                        break;
                    } else if (healths[last] < healths[idx]) {
                        healths[idx] -= 1;
                        healths[last] = -1;  // robot from stack is destroyed
                        stack.pop_back();
                    } else {
                        healths[idx] = -1;
                        healths[last] = -1;  // both robots are destroyed
                        stack.pop_back();
                        break;
                    }
                }
            }
        }
        
        vector<int> result;
        for (int i = 0; i < n; ++i) {
            if (healths[i] > 0) result.push_back(healths[i]);
        }
        return result;
    }
};

int main(){
    Solution s;
    int n;
    cin >> n;
    vector<int> positions(n);
    vector<int> healths(n);
    string directions;
    for (int i = 0; i < n; i++)
        cin >> positions[i];
    for (int i = 0; i < n; i++)
        cin >> healths[i];
    cin >> directions;
    vector<int> res = s.survivedRobotsHealths(positions, healths, directions);
    for (int i = 0; i < res.size(); i++)
        cout << res[i] << " ";
    cout << endl;
    return 0;

}